thewalrus/_hafnian.py

```diff
@@ -987,15 +931,75 @@ def hafnian_banded(A, loop=False, rtol=1e-05, atol=1e-08):
     return loop_haf[tuple(range(1, n + 1))]
 
 
+@numba.jit(nopython=True)
+def recursive_hafnian(A):  # pragma: no cover
+    r"""Computes the hafnian of the matrix with the recursive algorithm. It is an implementation of
+    algorithm 2 in *Counting perfect matchings as fast as Ryser* :cite:`bjorklund2012counting`.
+    Args:
+        A (array): the input matrix
+    Returns:
+        float: the hafnian of the input matrix
+    """
+    nb_lines, nb_columns = A.shape
+    if nb_lines != nb_columns:
+        raise ValueError("Matrix must be square")
+
+    if nb_lines % 2 != 0:
+        raise ValueError("Matrix size must be even")
+
+    n = int(float(len(A)) / 2)
+    z = np.zeros((n * (2 * n - 1), n + 1))
+    for j in range(1, 2 * n):
+        for k in range(j):
+            z[int(j * (j - 1) / 2 + k)][0] = A.copy()[j][k]
+    g = np.zeros(n + 1)
+    g[0] = 1
+    return solve(z, 2 * n, 1, g, n)
+
+
+@numba.jit(nopython=True)
+def solve(b, s, w, g, n):  # pragma: no cover
+    r"""Implements the recursive algorithm.
+    Args:
+        b (array): matrix that is transformed recursively
+        s (int): size of the original matrix that changes at every recursion
+        k (int): a variable of the recursive algorithm
+        g (int): matrix that is transformed recursively
+        n (int): size of the original matrix divided by 2
+    Returns:
+        float: the hafnian of the input matrix
+    """
+    if s == 0:
+        return w * g[n]
+    c = np.zeros((int((s - 2) * (s - 3) / 2), n + 1))
+    i = 0
+    for j in range(1, s - 2):
+        for k in range(j):
+            c[i] = b[int((j + 1) * (j + 2) / 2 + k + 2)]
+            i += 1
+    h = solve(c, s - 2, -w, g, n)
+    e = g[:].copy()
```

Comment:
DO you need to copy here?

Reply:
I think you do...

Reply:
Not sure, cause e gets changed iteratively, but g should stay the same

Reply:
We're afraid it might create memory issues

Reply:
Yeah, you are right, for this one I think you do...

Reply:
Why do you create a reference of the array first (using `g[:]`) and then a copy (with `.copy()`). To create a deep copy of the array `g.copy()` will suffice 🤓